src/BoundaryConditions/fill_halo_regions.jl

```diff
@@ -2,72 +2,79 @@
 ##### General halo filling functions
 #####
 
-@inline fill_halo_regions!(::Nothing, args...) = nothing
+fill_halo_regions!(::Nothing, args...) = []
 
 """
     fill_halo_regions!(fields, arch)
 
 Fill halo regions for each field in the tuple `fields` according to their boundary
 conditions, possibly recursing into `fields` if it is a nested tuple-of-tuples.
 """
-@inline function fill_halo_regions!(fields::Union{Tuple, NamedTuple}, arch, args...)
+function fill_halo_regions!(fields::Union{Tuple, NamedTuple}, arch, args...)
+
     for field in fields
         fill_halo_regions!(field, arch, args...)
     end
+
     return nothing
 end
 
-@inline fill_halo_regions!(field, arch, args...) =
+fill_halo_regions!(field, arch, args...) =
     fill_halo_regions!(field.data, field.boundary_conditions, arch, field.grid, args...)
 
 "Fill halo regions in x, y, and z for a given field."
-@inline function fill_halo_regions!(c::AbstractArray, fieldbcs, arch, args...)
-      fill_west_halo!(c, fieldbcs.x.left,   arch, args...)
-      fill_east_halo!(c, fieldbcs.x.right,  arch, args...)
-     fill_south_halo!(c, fieldbcs.y.left,   arch, args...)
-     fill_north_halo!(c, fieldbcs.y.right,  arch, args...)
-    fill_bottom_halo!(c, fieldbcs.z.bottom, arch, args...)
-       fill_top_halo!(c, fieldbcs.z.top,    arch, args...)
+function fill_halo_regions!(c::AbstractArray, fieldbcs, arch, grid, args...)
+
```

Comment:
This might be source of the race conditions. All of these kernels are missing a top synchronization,, thus it is possible for them to run earlier than they should possibly reading data that hasn't been updated.